// will-finance-client/src/components/ui/PhoenixLogo.tsx
import React from 'react';
import { motion } from 'framer-motion';

interface PhoenixLogoProps {
  readonly className?: string;
  readonly size?: 'sm' | 'md' | 'lg' | 'xl' | 'hero';
}

export function PhoenixLogo({ className, size = 'md' }: PhoenixLogoProps) {
  const sizeClasses = {
    sm: 'w-12 h-12',     // Aumentado de w-8 h-8
    md: 'w-16 h-16',     // Aumentado de w-12 h-12
    lg: 'w-24 h-24',     // Aumentado de w-16 h-16  
    xl: 'w-32 h-32',     // Aumentado de w-20 h-20
    hero: 'w-40 h-40'    // Novo tamanho para destaque
  };

  const finalSize = className || sizeClasses[size];

  // Extract border class for glow ring
  let glowRingBorderClass = '';
  if (size === 'hero') {
    glowRingBorderClass = 'border-2 border-orange-500/30';
  } else if (size === 'xl') {
    glowRingBorderClass = 'border-2 border-orange-500/25';
  } else {
    glowRingBorderClass = 'border border-orange-500/20';
  }

  return (
    <motion.div 
      className={`relative ${finalSize} flex-shrink-0 glass rounded-full border-2 border-cyber-primary shadow-[0_0_32px_4px_rgba(0,255,255,0.25)] animate-float ${className}`}
      initial={{ opacity: 0, scale: 0.8, rotate: -10 }}
      animate={{ opacity: 1, scale: 1, rotate: 0 }}
      transition={{ duration: 0.8, ease: "easeOut" }}
      style={{ background: 'linear-gradient(135deg, #0A0A2F 0%, #1A1A4E 100%)' }}
    >
      <motion.img
        src="/phoenix-logo.png"
        alt="Phoenix Logo"
        className="w-full h-full object-contain drop-shadow-cyber animate-glow"
        style={{ filter: 'drop-shadow(0 0 24px #00FFFF)', maxWidth: '100%', maxHeight: '100%' }}
        initial={{ opacity: 0 }}
        animate={{ opacity: 1 }}
        transition={{ duration: 1, delay: 0.2 }}
      />
      <motion.div
        className={`absolute inset-0 rounded-full ${glowRingBorderClass}`}
        animate={{ scale: [1, 1.15, 1], opacity: [0.3, 0.1, 0.3] }}
        transition={{ duration: 3, repeat: Infinity, ease: "easeInOut" }}
      />
    </motion.div>
  );
}
