// will-finance-client/src/pages/Budgets/BudgetsPage.tsx
import React, { useState } from 'react';
import { motion } from 'framer-motion';
import { 
  Plus, 
  Target, 
  TrendingUp, 
  AlertTriangle,
  CheckCircle,
  PiggyBank,
  Wallet,
  ShoppingCart,
  Home,
  Car,
  Coffee
} from 'lucide-react';
import { BudgetModal, BudgetData } from '../../components/Modal/BudgetModal';

interface Budget {
  id: string;
  name: string;
  category: string;
  allocated: number;
  spent: number;
  icon: React.ComponentType<{ className?: string }>;
  color: string;
  status: 'healthy' | 'warning' | 'danger';
}

const BudgetsPage = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [editingBudget, setEditingBudget] = useState<BudgetData | null>(null);
  
  const handleNewBudget = () => {
    setEditingBudget(null);
    setIsModalOpen(true);
  };

  const handleEditBudget = (budget: Budget) => {
    setEditingBudget({
      id: budget.id,
      name: budget.name,
      amount: budget.allocated,
      spent: budget.spent,
      category: budget.category,
      period: 'monthly',
      startDate: new Date().toISOString().split('T')[0],
      endDate: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000).toISOString().split('T')[0]
    });
    setIsModalOpen(true);
  };

  const handleSaveBudget = (budgetData: BudgetData) => {
    if (editingBudget) {
      // Atualizar orçamento existente
      setBudgets(prev => prev.map(b => 
        b.id === editingBudget.id 
          ? { 
              ...b, 
              name: budgetData.name,
              allocated: budgetData.amount,
              spent: budgetData.spent,
              category: budgetData.category
            }
          : b
      ));
    } else {
      // Criar novo orçamento
      const newBudget: Budget = {
        id: Date.now().toString(),
        name: budgetData.name,
        category: budgetData.category,
        allocated: budgetData.amount,
        spent: budgetData.spent,
        icon: ShoppingCart, // Default icon
        color: 'cyber-primary',
        status: budgetData.spent > budgetData.amount * 0.9 ? 'danger' : 
                budgetData.spent > budgetData.amount * 0.7 ? 'warning' : 'healthy'
      };
      setBudgets(prev => [...prev, newBudget]);
    }
    setIsModalOpen(false);
  };

  const [budgets, setBudgets] = useState<Budget[]>([]);

  // Busca dados reais do backend
  React.useEffect(() => {
    const fetchBudgets = async () => {
      try {
        const response = await fetch('/api/budgets', {
          headers: { 'Content-Type': 'application/json' },
        });
        if (!response.ok) throw new Error('Erro ao buscar orçamentos');
        const data = await response.json();
        setBudgets(data.budgets || []);
      } catch (err) {
        setBudgets([]);
      }
    };
    fetchBudgets();
  }, []);

  const totalAllocated = budgets.reduce((sum, budget) => sum + budget.allocated, 0);
  const totalSpent = budgets.reduce((sum, budget) => sum + budget.spent, 0);
  const remainingBudget = totalAllocated - totalSpent;

  const cardVariants = {
    hidden: { opacity: 0, y: 20 },
    visible: { opacity: 1, y: 0 }
  };

  const getProgressColor = (status: string) => {
    switch (status) {
      case 'healthy': return 'from-cyber-accent to-cyber-primary';
      case 'warning': return 'from-yellow-400 to-cyber-secondary';
      case 'danger': return 'from-cyber-danger to-red-500';
      default: return 'from-cyber-primary to-cyber-secondary';
    }
  };

  const getStatusIcon = (status: string) => {
    switch (status) {
      case 'healthy': return <CheckCircle className="w-5 h-5 text-cyber-accent" />;
      case 'warning': return <AlertTriangle className="w-5 h-5 text-yellow-400" />;
      case 'danger': return <AlertTriangle className="w-5 h-5 text-cyber-danger" />;
      default: return <Target className="w-5 h-5 text-cyber-primary" />;
    }
  };

  return (
    <div className="space-y-8">
      {/* Header */}
      <motion.div
        initial={{ opacity: 0, y: -20 }}
        animate={{ opacity: 1, y: 0 }}
        transition={{ duration: 0.6 }}
        className="flex flex-col lg:flex-row lg:items-center lg:justify-between"
      >
        <div>
          <h1 className="text-4xl font-cyber text-cyber-primary mb-2 text-glow">
            Orçamentos
          </h1>
          <p className="text-white-muted font-mono">
            Controle inteligente de gastos por categoria
          </p>
        </div>
        
        <motion.button
          whileHover={{ scale: 1.05 }}
          whileTap={{ scale: 0.95 }}
          onClick={handleNewBudget}
          className="btn-primary mt-4 lg:mt-0 self-start lg:self-auto"
        >
          <Plus className="w-5 h-5 mr-2" />
          <span className="font-mono">Novo Orçamento</span>
        </motion.button>
      </motion.div>

      {/* Summary Cards */}
      <div className="grid grid-cols-1 md:grid-cols-3 gap-6">
        <motion.div
          variants={cardVariants}
          initial="hidden"
          animate="visible"
          transition={{ duration: 0.6, delay: 0.1 }}
          className="glass p-6 rounded-xl border border-cyber-primary/30 hover:shadow-glow transition-all duration-300"
        >
          <div className="flex items-center justify-between mb-4">
            <div className="w-12 h-12 bg-gradient-cyber rounded-lg flex items-center justify-center">
              <Wallet className="w-6 h-6 text-background" />
            </div>
            <Target className="w-5 h-5 text-cyber-primary" />
          </div>
          <h3 className="text-sm font-mono text-white-muted mb-1">Orçamento Total</h3>
          <p className="text-2xl font-bold text-cyber-primary">
            R$ {totalAllocated.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
          </p>
        </motion.div>

        <motion.div
          variants={cardVariants}
          initial="hidden"
          animate="visible"
          transition={{ duration: 0.6, delay: 0.2 }}
          className="glass p-6 rounded-xl border border-cyber-danger/30 hover:shadow-[0_0_20px_#FF0040] transition-all duration-300"
        >
          <div className="flex items-center justify-between mb-4">
            <div className="w-12 h-12 bg-gradient-to-br from-cyber-danger to-cyber-secondary rounded-lg flex items-center justify-center">
              <TrendingUp className="w-6 h-6 text-background" />
            </div>
            <AlertTriangle className="w-5 h-5 text-cyber-danger" />
          </div>
          <h3 className="text-sm font-mono text-white-muted mb-1">Total Gasto</h3>
          <p className="text-2xl font-bold text-cyber-danger">
            R$ {totalSpent.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
          </p>
        </motion.div>

        <motion.div
          variants={cardVariants}
          initial="hidden"
          animate="visible"
          transition={{ duration: 0.6, delay: 0.3 }}
          className={`glass p-6 rounded-xl border transition-all duration-300 ${
            remainingBudget >= 0 
              ? 'border-cyber-accent/30 hover:shadow-[0_0_20px_#39FF14]' 
              : 'border-cyber-danger/30 hover:shadow-[0_0_20px_#FF0040]'
          }`}
        >
          <div className="flex items-center justify-between mb-4">
            <div className={`w-12 h-12 bg-gradient-to-br rounded-lg flex items-center justify-center ${
              remainingBudget >= 0 
                ? 'from-cyber-accent to-cyber-primary' 
                : 'from-cyber-danger to-cyber-secondary'
            }`}>
              <PiggyBank className="w-6 h-6 text-background" />
            </div>
            {remainingBudget >= 0 ? (
              <CheckCircle className="w-5 h-5 text-cyber-accent" />
            ) : (
              <AlertTriangle className="w-5 h-5 text-cyber-danger" />
            )}
          </div>
          <h3 className="text-sm font-mono text-white-muted mb-1">Saldo Restante</h3>
          <p className={`text-2xl font-bold ${
            remainingBudget >= 0 ? 'text-cyber-accent' : 'text-cyber-danger'
          }`}>
            R$ {Math.abs(remainingBudget).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
          </p>
        </motion.div>
      </div>

      {/* Budget Progress Overview */}
      <motion.div
        variants={cardVariants}
        initial="hidden"
        animate="visible"
        transition={{ duration: 0.6, delay: 0.4 }}
        className="glass p-6 rounded-xl border border-cyber-primary/20"
      >
        <h3 className="text-lg font-cyber text-cyber-primary mb-6">Progresso Geral</h3>
        
        <div className="space-y-4">
          <div className="flex justify-between text-sm font-mono">
            <span className="text-white-muted">Utilização do Orçamento</span>
            <span className="text-cyber-primary">{((totalSpent / totalAllocated) * 100).toFixed(1)}%</span>
          </div>
          
          <div className="w-full bg-black-tertiary rounded-full h-4 overflow-hidden">
            <motion.div
              initial={{ width: 0 }}
              animate={{ width: `${Math.min((totalSpent / totalAllocated) * 100, 100)}%` }}
              transition={{ duration: 1.5, delay: 1 }}
              className={`h-full bg-gradient-to-r relative ${
                totalSpent <= totalAllocated * 0.8 
                  ? 'from-cyber-accent to-cyber-primary'
                  : totalSpent <= totalAllocated
                  ? 'from-yellow-400 to-cyber-secondary'
                  : 'from-cyber-danger to-red-500'
              }`}
            >
              <div className="absolute inset-0 bg-gradient-to-r from-transparent via-white/20 to-transparent animate-shimmer"></div>
            </motion.div>
          </div>
          
          <div className="flex justify-between text-sm font-mono">
            <span className="text-white-secondary">
              R$ {totalSpent.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
            </span>
            <span className="text-cyber-primary">
              R$ {totalAllocated.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
            </span>
          </div>
        </div>
      </motion.div>

      {/* Budget Categories */}
      <div className="grid grid-cols-1 lg:grid-cols-2 gap-6">
        {budgets.map((budget, index) => {
          const percentage = (budget.spent / budget.allocated) * 100;
          const Icon = budget.icon;
          
          return (
            <motion.div
              key={budget.id}
              variants={cardVariants}
              initial="hidden"
              animate="visible"
              transition={{ duration: 0.6, delay: 0.5 + index * 0.1 }}
              className="glass p-6 rounded-xl border border-cyber-primary/20 hover:border-cyber-primary/40 transition-all duration-300 group cursor-pointer"
              onClick={() => handleEditBudget(budget)}
            >
              <div className="flex items-center justify-between mb-4">
                <div className="flex items-center space-x-3">
                  <div className={`w-12 h-12 bg-gradient-to-br from-${budget.color} to-cyber-primary rounded-lg flex items-center justify-center`}>
                    <Icon className="w-6 h-6 text-background" />
                  </div>
                  <div>
                    <h4 className="font-medium text-white group-hover:text-cyber-primary transition-colors">
                      {budget.name}
                    </h4>
                    <p className="text-sm text-white-muted font-mono">{budget.category}</p>
                  </div>
                </div>
                {getStatusIcon(budget.status)}
              </div>

              <div className="space-y-3">
                <div className="flex justify-between text-sm font-mono">
                  <span className="text-white-muted">Progresso</span>
                  <span className={`${percentage > 100 ? 'text-cyber-danger' : 'text-cyber-primary'}`}>
                    {percentage.toFixed(1)}%
                  </span>
                </div>
                
                <div className="w-full bg-black-tertiary rounded-full h-2 overflow-hidden">
                  <motion.div
                    initial={{ width: 0 }}
                    animate={{ width: `${Math.min(percentage, 100)}%` }}
                    transition={{ duration: 1, delay: 1 + index * 0.1 }}
                    className={`h-full bg-gradient-to-r ${getProgressColor(budget.status)} relative`}
                  >
                    <div className="absolute inset-0 bg-gradient-to-r from-transparent via-white/20 to-transparent animate-shimmer"></div>
                  </motion.div>
                </div>
                
                <div className="flex justify-between text-sm font-mono">
                  <span className="text-white-secondary">
                    R$ {budget.spent.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
                  </span>
                  <span className="text-white-muted">
                    de R$ {budget.allocated.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
                  </span>
                </div>

                {percentage > 80 && (
                  <div className={`flex items-center space-x-2 text-xs font-mono p-2 rounded ${
                    percentage > 100 
                      ? 'bg-cyber-danger/20 text-cyber-danger' 
                      : 'bg-yellow-400/20 text-yellow-400'
                  }`}>
                    <AlertTriangle className="w-4 h-4" />
                    <span>
                      {percentage > 100 
                        ? `Orçamento excedido em R$ ${(budget.spent - budget.allocated).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`
                        : 'Próximo do limite do orçamento'
                      }
                    </span>
                  </div>
                )}
              </div>
            </motion.div>
          );
        })}
      </div>
      {/* Budget Modal */}
      <BudgetModal
        isOpen={isModalOpen}
        onClose={() => setIsModalOpen(false)}
        onSave={handleSaveBudget}
        budget={editingBudget}
      />
    </div>
  );
}
export default BudgetsPage;
