// server/src/controllers/dashboard.controller.ts
import { Request, Response, NextFunction } from 'express';
import { PrismaClient } from '@prisma/client';
const prisma = new PrismaClient();

export async function getDashboardStats(
  req: Request,
  res: Response,
  next: NextFunction,
) {
  try {
    // Exemplo: buscar receitas, despesas, saldo, total de transações
    const userId = req.user?.id || 'user-1'; // mock para teste
    const income = await prisma.transaction.aggregate({
      _sum: { amount: true },
      where: { userId, type: 'INCOME' },
    });
    const expense = await prisma.transaction.aggregate({
      _sum: { amount: true },
      where: { userId, type: 'EXPENSE' },
    });
    const total = await prisma.transaction.count({ where: { userId } });
    res.json({
      success: true,
      data: {
        income: income._sum.amount || 0,
        expense: expense._sum.amount || 0,
        balance: (income._sum.amount || 0) - (expense._sum.amount || 0),
        total,
      },
    });
  } catch (error) {
    next(error);
  }
}
