// will-finance-client/src/pages/Transactions/TransactionsPage.tsx
import { useEffect, useState, useCallback, useMemo } from 'react';
import { motion } from 'framer-motion';
import { showSuccess, showError } from '../../stores/uiStore';
import { useAuthStore } from '../../stores/authStore';
import { useMasterUser } from '../../hooks/useMasterUser';
import { 
  Plus, 
  Search, 
  Filter, 
  Download, 
  Upload, 
  CreditCard,
  TrendingUp,
  TrendingDown,
  Calendar,
  DollarSign
} from 'lucide-react';

// Types para melhor type safety
interface Transaction {
  id: string;
  description: string;
  amount: number;
  type: 'INCOME' | 'EXPENSE';
  category: string;
  date: string;
  account?: string;
}

interface TransactionFilters {
  search: string;
  type: 'ALL' | 'INCOME' | 'EXPENSE';
  category: string;
  dateFrom: string;
  dateTo: string;
}

// Utility function melhorada
function getUserDisplayName(user?: { name?: string; email?: string; displayName?: string }) {
  if (!user) return 'Usuário';
  return user.displayName || user.name || user.email?.split('@')[0] || 'Usuário';
}


const TransactionsPage = () => {
  const { user, token } = useAuthStore();
  const { isMaster, databases } = useMasterUser();
  
  // Estados otimizados
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [transactions, setTransactions] = useState<Transaction[]>([]);
  const [filters, setFilters] = useState<TransactionFilters>({
    search: '',
    type: 'ALL',
    category: '',
    dateFrom: '',
    dateTo: ''
  });

  // Fetch transactions com dados reais do banco
  const fetchTransactions = useCallback(async () => {
    try {
      setLoading(true);
      setError(null);

      const response = await fetch('/api/transactions', {
        headers: {
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        throw new Error(`Erro ${response.status}: ${response.statusText}`);
      }

      const data = await response.json();
      setTransactions(data.transactions || []);

      // Mostrar toast apenas na primeira carga
      if (transactions.length === 0) {
        const welcomeMessage = isMaster
          ? `Bem-vindo! Acesso Master ativado.`
          : `${data.transactions?.length || 0} transações carregadas.`;
        showSuccess(welcomeMessage);
      }

    } catch (err) {
      console.error('Erro ao carregar transações:', err);
      const errorMessage = err instanceof Error ? err.message : 'Erro ao carregar transações';
      setError(errorMessage);
      setTransactions([]);
      
      // Mostrar toast de erro apenas se não for problema de rede comum
      if (!errorMessage.includes('Failed to fetch')) {
  showError(errorMessage);
      }
    } finally {
      setLoading(false);
    }
  }, [isMaster, databases, transactions.length]);

  // Effect otimizado para evitar loops - carrega apenas uma vez
  useEffect(() => {
    let isMounted = true;
    let hasLoaded = false;
    
    const loadTransactions = async () => {
      if (isMounted && !hasLoaded) {
        hasLoaded = true;
        await fetchTransactions();
      }
    };

    loadTransactions();
    
    return () => {
      isMounted = false;
    };
  }, []); // Dependências removidas para evitar loops

  // Transações filtradas
  const filteredTransactions = useMemo(() => {
    return transactions.filter(transaction => {
      const matchesSearch = !filters.search || 
        transaction.description.toLowerCase().includes(filters.search.toLowerCase()) ||
        transaction.category.toLowerCase().includes(filters.search.toLowerCase());
      
      const matchesType = filters.type === 'ALL' || transaction.type === filters.type;
      
      const matchesCategory = !filters.category || transaction.category === filters.category;
      
      const transactionDate = new Date(transaction.date);
      const matchesDateFrom = !filters.dateFrom || transactionDate >= new Date(filters.dateFrom);
      const matchesDateTo = !filters.dateTo || transactionDate <= new Date(filters.dateTo);
      
      return matchesSearch && matchesType && matchesCategory && matchesDateFrom && matchesDateTo;
    });
  }, [transactions, filters]);

  // Estatísticas calculadas
  const stats = useMemo(() => {
    const income = filteredTransactions
      .filter(t => t.type === 'INCOME')
      .reduce((sum, t) => sum + Math.abs(t.amount), 0);
    
    const expenses = filteredTransactions
      .filter(t => t.type === 'EXPENSE')  
      .reduce((sum, t) => sum + Math.abs(t.amount), 0);
    
    return {
      income,
      expenses,
      balance: income - expenses,
      total: filteredTransactions.length
    };
  }, [filteredTransactions]);

  // Formatação de moeda
  const formatCurrency = (amount: number) => {
    return new Intl.NumberFormat('pt-BR', {
      style: 'currency',
      currency: 'BRL'
    }).format(Math.abs(amount));
  };

  // Formatação de data
  const formatDate = (dateString: string) => {
    return new Date(dateString).toLocaleDateString('pt-BR', {
      day: '2-digit',
      month: '2-digit',
      year: 'numeric',
      hour: '2-digit',
      minute: '2-digit'
    });
  };

  return (
    <motion.div
      initial={{ opacity: 0, y: 20 }}
      animate={{ opacity: 1, y: 0 }}
      exit={{ opacity: 0, y: -20 }}
      transition={{ duration: 0.3 }}
      className="transactions-page p-6"
    >
      <div className="max-w-7xl mx-auto">
        {/* Header */}
        <header className="mb-8">
          <h1 className="text-3xl font-bold text-cyber-primary flex items-center gap-3">
            <CreditCard className="w-8 h-8" />
            Transações
          </h1>
          <p className="text-white-muted mt-2">
            Gerenciamento Financeiro Avançado
          </p>
        </header>

        {/* Stats Cards */}
          <div className="grid grid-cols-1 md:grid-cols-4 gap-6 mb-8">
            {/* Cards modernizados cyberpunk */}
          <motion.div 
            className="glass p-6 rounded-lg"
            whileHover={{ scale: 1.02 }}
          >
            <div className="flex items-center justify-between">
              <div>
                <p className="text-white-muted text-sm">Receitas</p>
                <p className="text-2xl font-bold text-green-400">
                  {formatCurrency(stats.income)}
                </p>
              </div>
              <TrendingUp className="w-8 h-8 text-green-400" />
            </div>
          </motion.div>

          <motion.div 
            className="glass p-6 rounded-lg"
            whileHover={{ scale: 1.02 }}
          >
            <div className="flex items-center justify-between">
              <div>
                <p className="text-white-muted text-sm">Despesas</p>
                <p className="text-2xl font-bold text-red-400">
                  {formatCurrency(stats.expenses)}
                </p>
              </div>
              <TrendingDown className="w-8 h-8 text-red-400" />
            </div>
          </motion.div>

          <motion.div 
            className="glass p-6 rounded-lg"
            whileHover={{ scale: 1.02 }}
          >
            <div className="flex items-center justify-between">
              <div>
                <p className="text-white-muted text-sm">Saldo</p>
                <p className={`text-2xl font-bold ${stats.balance >= 0 ? 'text-cyber-primary' : 'text-red-400'}`}>
                  {formatCurrency(stats.balance)}
                </p>
              </div>
              <DollarSign className="w-8 h-8 text-cyber-primary" />
            </div>
          </motion.div>

          <motion.div 
            className="glass p-6 rounded-lg"
            whileHover={{ scale: 1.02 }}
          >
            <div className="flex items-center justify-between">
              <div>
                <p className="text-white-muted text-sm">Total</p>
                <p className="text-2xl font-bold text-cyber-primary">
                  {stats.total}
                </p>
              </div>
              <Calendar className="w-8 h-8 text-cyber-primary" />
            </div>
          </motion.div>
        </div>

        {/* Actions Bar */}
        <div className="flex flex-col md:flex-row gap-4 mb-6">
          <div className="flex-1">
            <div className="relative">
              <Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-white-muted w-4 h-4" />
              <input
                type="text"
                placeholder="Buscar transações..."
                value={filters.search}
                onChange={(e) => setFilters((prev: typeof filters) => ({ ...prev, search: e.target.value }))}
                className="input pl-10 w-full"
              />
            </div>
          </div>
          
          <div className="flex gap-3">
            <button className="btn btn-secondary flex items-center gap-2">
              <Filter className="w-4 h-4" />
              Filtros
            </button>
            <button className="btn btn-secondary flex items-center gap-2">
              <Upload className="w-4 h-4" />
              Importar
            </button>
            <button className="btn btn-secondary flex items-center gap-2">
              <Download className="w-4 h-4" />
              Exportar
            </button>
            <button className="btn btn-primary flex items-center gap-2">
              <Plus className="w-4 h-4" />
              Nova Transação
            </button>
          </div>
        </div>

        {/* Content */}
        <div className="glass rounded-lg overflow-hidden">
          {loading ? (
            <div className="flex items-center justify-center p-12">
              <div className="animate-spin rounded-full h-8 w-8 border-b-2 border-cyber-primary"></div>
              <span className="ml-3 text-white-muted">Carregando transações...</span>
            </div>
          ) : error ? (
            <div className="p-8 text-center">
              <div className="text-red-400 mb-4">⚠️ {error}</div>
              <button 
                onClick={fetchTransactions}
                className="btn btn-primary"
              >
                Tentar Novamente
              </button>
            </div>
          ) : filteredTransactions.length === 0 ? (
            <div className="p-12 text-center">
              <CreditCard className="w-16 h-16 text-white-muted mx-auto mb-4" />
              <h3 className="text-lg font-medium text-white-primary mb-2">
                Nenhuma Transação Encontrada
              </h3>
              <p className="text-white-muted mb-6">
                Comece adicionando sua primeira transação
              </p>
              <button className="btn btn-primary">
                <Plus className="w-4 h-4 mr-2" />
                Adicionar Transação
              </button>
            </div>
          ) : (
            <div className="overflow-x-auto">
              <table className="w-full">
                <thead className="bg-black-secondary/50">
                  <tr>
                    <th className="text-left p-4 text-white-secondary font-medium">Data</th>
                    <th className="text-left p-4 text-white-secondary font-medium">Descrição</th>
                    <th className="text-left p-4 text-white-secondary font-medium">Categoria</th>
                    <th className="text-left p-4 text-white-secondary font-medium">Conta</th>
                    <th className="text-right p-4 text-white-secondary font-medium">Valor</th>
                  </tr>
                </thead>
                <tbody>
                  {filteredTransactions.map((transaction, index: number) => (
                    <motion.tr
                      key={transaction.id}
                      initial={{ opacity: 0, y: 10 }}
                      animate={{ opacity: 1, y: 0 }}
                      transition={{ delay: index * 0.05 }}
                      className="border-t border-cyber-border-muted hover:bg-black-secondary/30 transition-colors"
                    >
                      <td className="p-4 text-white-muted text-sm">
                        {formatDate(transaction.date)}
                      </td>
                      <td className="p-4">
                        <div className="font-medium text-white-primary">
                          {transaction.description}
                        </div>
                      </td>
                      <td className="p-4">
                        <span className="inline-flex items-center px-2 py-1 rounded-full text-xs font-medium bg-cyber-primary/10 text-cyber-primary">
                          {transaction.category}
                        </span>
                      </td>
                      <td className="p-4 text-white-muted text-sm">
                        {transaction.account || 'N/A'}
                      </td>
                      <td className="p-4 text-right">
                        <span className={`font-bold ${
                          transaction.type === 'INCOME' 
                            ? 'text-green-400' 
                            : 'text-red-400'
                        }`}>
                          {transaction.type === 'INCOME' ? '+' : '-'}
                          {formatCurrency(transaction.amount)}
                        </span>
                      </td>
                    </motion.tr>
                  ))}
                </tbody>
              </table>
            </div>
          )}
        </div>

        {/* Status Bar */}
        <div className="mt-6 flex justify-between items-center text-sm text-white-muted">
          <div>
            Sistema Online • {formatDate(new Date().toISOString())}
          </div>
          <div>
            {isMaster && (
              <span className="text-cyber-primary font-medium">
                Modo Master Ativo • Bancos: {databases.join(', ')}
              </span>
            )}
          </div>
        </div>
      </div>
    </motion.div>
  );
}
export default TransactionsPage;
