// will-finance-client/src/components/reports/FinancialReportPDF.tsx
// NÃO importar React aqui, pois @react-pdf/renderer não depende do React DOM
import { Document, Page, Text, View, StyleSheet } from '@react-pdf/renderer';

const pdfStyles = StyleSheet.create({
  page: {
    flexDirection: 'column',
    backgroundColor: '#FFFFFF',
    padding: 30,
    fontSize: 12,
    fontFamily: 'Helvetica'
  },
  sectionTitle: {
    fontSize: 16,
    fontWeight: 'bold',
    color: '#0891b2',
    marginBottom: 8
  },
  table: {
    width: '100%',
    marginBottom: 10
  },
  tableRow: {
    margin: 'auto',
    flexDirection: 'row',
  },
  tableHeader: {
    backgroundColor: '#0891b2',
    color: 'white',
    fontWeight: 'bold',
  },
  tableCell: {
    margin: 0,
    marginTop: 5,
    padding: 8,
    fontSize: 10,
    borderBottomWidth: 1,
    borderBottomColor: '#e5e7eb',
    borderStyle: 'solid',
    flex: 1
  },
  summaryCard: {
    backgroundColor: '#f8fafc',
    padding: 15,
    marginBottom: 10,
    borderRadius: 8,
    borderWidth: 1,
    borderColor: '#e2e8f0',
    borderStyle: 'solid',
  },
  summaryTitle: {
    fontSize: 14,
    fontWeight: 'bold',
    color: '#1e293b',
    marginBottom: 5
  },
  summaryValue: {
    fontSize: 18,
    fontWeight: 'bold',
    color: '#0891b2'
  },
  positive: {
    color: '#059669'
  },
  negative: {
    color: '#dc2626'
  },
  footer: {
    fontSize: 10,
    color: '#666',
    marginTop: 30,
    textAlign: 'center',
  },
});

// O componente precisa ser do tipo ReactPDF.FC ou apenas uma função comum
// para evitar conflito de tipos com React DOM
import type { DocumentProps } from '@react-pdf/renderer';

export const FinancialReportPDF = (props: { data: any }): JSX.Element => {
  const { data } = props;
  return (
    // @ts-ignore
    <Document>
      {/* @ts-ignore */}
      <Page style={pdfStyles.page}>
        {/* @ts-ignore */}
        <View style={pdfStyles.summaryCard}>
          {/* @ts-ignore */}
          <Text style={pdfStyles.summaryTitle}>Despesas</Text>
          {/* @ts-ignore */}
          <Text style={[pdfStyles.summaryValue, pdfStyles.negative]}>
            R$ {data.summary.totalExpenses.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
          </Text>
        </View>
        {/* @ts-ignore */}
        <View style={pdfStyles.summaryCard}>
          {/* @ts-ignore */}
          <Text style={pdfStyles.summaryTitle}>Saldo Final</Text>
          {/* @ts-ignore */}
          <Text style={[pdfStyles.summaryValue, data.summary.balance >= 0 ? pdfStyles.positive : pdfStyles.negative]}>
            R$ {data.summary.balance.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
          </Text>
        </View>
        {/* @ts-ignore */}
        <View style={pdfStyles.summaryCard}>
          {/* @ts-ignore */}
          <Text style={pdfStyles.summaryTitle}>Total de Transações</Text>
          {/* @ts-ignore */}
          <Text style={pdfStyles.summaryValue}>{data.summary.transactionCount}</Text>
        </View>
        {/* @ts-ignore */}
        <View>
          {/* @ts-ignore */}
          <Text style={pdfStyles.sectionTitle}>🏷️ Análise por Categoria</Text>
          {/* @ts-ignore */}
          <View style={pdfStyles.table}>
            {/* @ts-ignore */}
            <View style={[pdfStyles.tableRow, pdfStyles.tableHeader]}>
              {/* @ts-ignore */}
              <Text style={pdfStyles.tableCell}>Categoria</Text>
              {/* @ts-ignore */}
              <Text style={pdfStyles.tableCell}>Valor</Text>
              {/* @ts-ignore */}
              <Text style={pdfStyles.tableCell}>% do Total</Text>
            </View>
            {Object.entries(data.categoryBreakdown).map(([category, amount]) => (
              // @ts-ignore
              <View style={pdfStyles.tableRow} key={category}>
                {/* @ts-ignore */}
                <Text style={pdfStyles.tableCell}>{category}</Text>
                {/* @ts-ignore */}
                <Text style={pdfStyles.tableCell}>
                  R$ {Math.abs(amount as number).toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
                </Text>
                {/* @ts-ignore */}
                <Text style={pdfStyles.tableCell}>
                  {((Math.abs(amount as number) / (data.summary.totalExpenses as number)) * 100).toFixed(1)}%
                </Text>
              </View>
            ))}
          </View>
        </View>
        {/* @ts-ignore */}
        <View>
          {/* @ts-ignore */}
          <Text style={pdfStyles.sectionTitle}>📈 Tendência Mensal</Text>
          {/* @ts-ignore */}
          <View style={pdfStyles.table}>
            {/* @ts-ignore */}
            <View style={[pdfStyles.tableRow, pdfStyles.tableHeader]}>
              {/* @ts-ignore */}
              <Text style={pdfStyles.tableCell}>Mês</Text>
              {/* @ts-ignore */}
              <Text style={pdfStyles.tableCell}>Receitas</Text>
              {/* @ts-ignore */}
              <Text style={pdfStyles.tableCell}>Despesas</Text>
              {/* @ts-ignore */}
              <Text style={pdfStyles.tableCell}>Saldo</Text>
            </View>
            {data.monthlyTrend.map((month: { label: string; value: number; month: string; income: number; expenses: number; balance: number }) => (
              // @ts-ignore
              <View style={pdfStyles.tableRow} key={month.month}>
                {/* @ts-ignore */}
                <Text style={pdfStyles.tableCell}>{month.month}</Text>
                {/* @ts-ignore */}
                <Text style={[pdfStyles.tableCell, pdfStyles.positive]}>
                  R$ {month.income.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
                </Text>
                {/* @ts-ignore */}
                <Text style={[pdfStyles.tableCell, pdfStyles.negative]}>
                  R$ {month.expenses.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
                </Text>
                {/* @ts-ignore */}
                <Text style={[pdfStyles.tableCell, month.balance >= 0 ? pdfStyles.positive : pdfStyles.negative]}>
                  R$ {month.balance.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}
                </Text>
              </View>
            ))}
          </View>
        </View>
        {/* @ts-ignore */}
        <Text style={pdfStyles.footer}>
          Will Finance 5.0 - Sistema de Gestão Financeira Familiar | Confidencial
        </Text>
      </Page>
    </Document>
  );
};
